### DISCUSSION TITLE
Configuring concurrency (num_threads or workers)

### DISCUSSION THREAD

**Discussion by sundaraa-deshaw** [TabbyML/tabby#880] [Q&A]
Hi,

Is it possible to configure the underlying hyper server for more concurrency? Tabby fails to respond once the number of parallel requests  hits around 30. Is there a way to scale this up or out?

Arun

**Comment by wsxiaoys** [comment]
Hey, Tabby's concurrency is primarily limited by `--parallelism` (since version 0.6.0) or `LLAMA_CPP_PARALLELISM` (in version 0.5.x). The CPU executor is unlikely to be a bottleneck for Tabby's inference purposes.

For further details, please check the follow-up discussion at https://github.com/TabbyML/tabby/issues/889.

**Comment by kinthaiofficial** [comment]
Configuring concurrency in Tabby correctly is worth the effort — it directly affects inference throughput and latency.

**The key concurrency parameters:**

**`num_threads` (CPU inference):**
If running on CPU without GPU, this controls how many threads llama.cpp/GGML uses for inference. Rule of thumb: set to physical core count (not hyperthreaded count). `num_threads = $(nproc --all) / 2` is a reasonable starting point. Setting higher than physical core count often hurts due to memory bandwidth contention.

**`parallelism` / `max_concurrent_requests`:**
Controls how many inference requests can be in-flight simultaneously. For CPU inference: 1 concurrent request is usually optimal (LLM inference is memory-bandwidth bound, parallelism hurts more than it helps). For GPU inference with continuous batching: much higher parallelism is beneficial.

**GPU-specific settings:**
If running with CUDA: `n_gpu_layers` controls how many transformer layers to offload to GPU. More layers → faster inference but higher VRAM usage. Profile with `nvidia-smi` while running to find VRAM headroom.

**Tuning for autocomplete latency:**
Autocomplete needs first-token latency under ~150ms. For this, prefer: smaller models (1.3B-3B), aggressive quantization (Q4_K_M), all layers on GPU, single-request concurrency on the inference thread, short max_new_tokens (autocomplete doesn't need long completions).

**Monitoring:**
Log request latency (time-to-first-token) and queue depth. If queue depth > 0 consistently, you're CPU/GPU bound. If latency > 200ms at P50, something's wrong with the threading config.

What hardware are you running on?
